src/protocol/client.c

```diff
@@ -44,8 +38,246 @@ struct ProtocolClient
     const String *errorPrefix;                                      // Prefix used when throwing error
     TimeMSec keepAliveTime;                                         // Last time data was put to the server
+    uint64_t sessionTotal;                                          // Total sessions (used to generate session ids)
+    List *sessionList;                                              // Session list
+};
+
+struct ProtocolClientSession
+{
+    ProtocolClientSessionPub pub;                                   // Publicly accessible variables
+    ProtocolClient *client;                                         // Protocol client
+    StringId command;                                               // Command
+    uint64_t sessionId;                                             // Session id
+    bool async;                                                     // Async requests allowed?
+
+    // Stored message (read by another session and stored for later retrieval)
+    bool stored;                                                    // Is a message currently stored?
+    bool close;                                                     // Should the session be closed?
+    ProtocolMessageType type;                                       // Type of last message received
+    PackRead *packRead;                                             // Last message received (if any)
 };
 
+/***********************************************************************************************************************************
+Check protocol state
+***********************************************************************************************************************************/
+static void
+protocolClientStateExpectIdle(const ProtocolClient *const this)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PROTOCOL_CLIENT, this);
+    FUNCTION_TEST_END();
+
+    if (this->state != protocolClientStateIdle)
+    {
+        THROW_FMT(
+            ProtocolError, "client state is '%s' but expected '%s'", strZ(strIdToStr(this->state)),
+            strZ(strIdToStr(protocolClientStateIdle)));
+    }
+
+    FUNCTION_TEST_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Find a client session
+***********************************************************************************************************************************/
+static unsigned int
+protocolClientSessionFindIdx(const ProtocolClient *const this, const uint64_t sessionId)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PROTOCOL_CLIENT, this);
+        FUNCTION_TEST_PARAM(UINT64, sessionId);
+    FUNCTION_TEST_END();
+
+    unsigned int result = 0;
+
+    for (; result < lstSize(this->sessionList); result++)
+    {
+        if ((*(ProtocolClientSession **)lstGet(this->sessionList, result))->sessionId == sessionId)
```

Comment:
Hmm, well this is already encapsulated in protocolClientSessionFindIdx() so not sure encapsulating that single line in yet another function would add any clarity since this is only done once for this data type.

But in generally this pattern `(ProtocolClientSession **)lstGet(this->sessionList, result)` is not uncommon in the code base (search `**)lstGet(`). I think it's just the nature of C code.

Reply:
That's fine, something needs to be the low-level and if that's here makes sense.